cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/BasicFlowletContext.java

```diff
@@ -83,6 +85,18 @@ public MetricsCollector load(String key) throws Exception {
         }
       });
 
+    this.producerMetrics = CacheBuilder.newBuilder()
+      .expireAfterAccess(1, TimeUnit.HOURS)
+      .build(new CacheLoader<ImmutablePair<String, String>, MetricsCollector>() {
+        @Override
+        public MetricsCollector load(ImmutablePair<String, String> key) throws Exception {
+          return getProgramMetrics()
+            .childCollector(Constants.Metrics.Tag.QUEUE_PRODUCER, key.getFirst())
+            .childCollector(Constants.Metrics.Tag.FLOWLET_QUEUE, key.getSecond())
+            .childCollector(Constants.Metrics.Tag.QUEUE_CONSUMER, BasicFlowletContext.this.flowletId);
```

Comment:
childCollector accepts a Map<String, String> , its better to use that than multiple calls , as each of the childCollector call adds a map to the collectorCache.